### ISSUE TITLE
Library not loaded in mobile Safari

### ISSUE THREAD

**Issue by vcrepin** [tsayen/dom-to-image#18]
Hello. I'm trying to use domtoimage in Mobile Safari.

The library is not loaded and no error is apparent. I added an alert at the beginning of the script and it does not show. If I remove all the content of the script the alert shows.

I will investigate the issue but I was wondering if the library had been tested on safari.

**Comment by tsayen** [comment]
Hi! So far the library had been tested in FF and Chrome only, on desktop (unit tests run in those two browsers). I'm on Ubuntu and unfortunately desktop Safari is not available for Ubuntu. I wonder what could be done about that...

**Comment by vcrepin** [comment]
Hi!

I will investigate the issue and keep you posted.

**Comment by vcrepin** [comment]
By the way it works perfectly on mobile chrome and other android browsers.

**Comment by tsayen** [comment]
Nice to hear about mobile browsers, thanks for testing!

**Comment by vcrepin** [comment]
Hi. Just as a starter, I validated your script against Lint and it returns numerous warnings.

I invite you to post your script in http://www.javascriptlint.com/online_lint.php to view that.

**Comment by vcrepin** [comment]
This regex is not valid:`

/\.([^\./]*?)$/g

as shown by https://regex101.com/

what do you want to match with it ?

**Comment by vcrepin** [comment]
This is a 'lintified' version of your script. I will test it on safari tonight.

(function(global) {
    'use strict';

    var util = newUtil();
    var inliner = newInliner();
    var fontFaces = newFontFaces();
    var images = newImages();

    global.domtoimage = {
        toSvg: toSvg,
        toPng: toPng,
        toBlob: toBlob,
        impl: {
            fontFaces: fontFaces,
            images: images,
            util: util,
            inliner: inliner
        }
    };

    /**
     * @param {Node} node - The DOM Node object to render
     * @param {Object} options - Rendering options
     * @param {Function} options.filter - Should return true if passed node should be included in the output
     *          (excluding node means excluding it's children as well)
     * @param {String} options.bgcolor - color for the background, any valid CSS color value
     * @return {Promise} - A promise that is fulfilled with a SVG image data URL
     * */
    function toSvg(node, options) {
        options = options || {};
        return Promise.resolve(node).then(function(node) {
            return cloneNode(node, options.filter);
        }).then(embedFonts).then(inlineImages).then(function(clone) {
            if (options.bgcolor) {
                clone.style.backgroundColor = options.bgcolor;
            }
            return clone;
        }).then(function(clone) {
            return makeSvgDataUri(clone, node.scrollWidth, node.scrollHeight);
        });
    }

    /**
     * @param {Node} node - The DOM Node object to render
     * @param {Object} options - Rendering options, @see {@link toSvg}
     * @return {Promise} - A promise that is fulfilled with a PNG image data URL
     * */
    function toPng(node, options) {
        return draw(node, options || {}).then(function(canvas) {
            return canvas.toDataURL();
        });
    }

    /**
     * @param {Node} node - The DOM Node object to render
     * @param {Object} options - Rendering options, @see {@link toSvg}
     * @return {Promise} - A promise that is fulfilled with a PNG image blob
     * */
    function toBlob(node, options) {
        return draw(node, options || {}).then(util.canvasToBlob);
    }

    function cloneNode(node, filter) {
        if (filter && !filter(node)) {
            return Promise.resolve();
        }

        function cloneChildren(original, clone, filter) {
            var children = original.childNodes;
            if (children.length === 0) {
                return Promise.resolve(clone);
            }

            function cloneChildrenInOrder(parent, children, filter) {
                var done = Promise.resolve();
                children.forEach(function(child) {
                    done = done.then(function() {
                        return cloneNode(child, filter);
                    }).then(function(childClone) {
                        if (childClone) {
                            parent.appendChild(childClone);
                        }
                    });
                });
                return done;
            }

            return cloneChildrenInOrder(clone, util.asArray(children), filter).then(function() {
                return clone;
            });

        }

        function processClone(original, clone) {
            if (!(clone instanceof Element)) {
                return clone;
            }

            function cloneStyle() {
                copyStyle(global.window.getComputedStyle(original), clone.style);

                function copyStyle(source, target) {
                    if (source.cssText) {
                        target.cssText = source.cssText;
                    } else {
                        copyProperties(source, target);
                    }

                    function copyProperties(source, target) {
                        util.asArray(source).forEach(function(name) {
                            target.setProperty(name, source.getPropertyValue(name), source.getPropertyPriority(name));
                        });
                    }
                }
            }

            function clonePseudoElements() {
                [':before', ':after'].forEach(function(element) {
                    clonePseudoElement(element);
                });

                function clonePseudoElement(element) {
                    var style = global.window.getComputedStyle(original, element);
                    var content = style.getPropertyValue('content');

                    if (content === '' || content === 'none') {
                        return;
                    }

                    var className = util.uid();
                    clone.className = clone.className + ' ' + className;
                    var styleElement = global.document.createElement('style');
                    styleElement.appendChild(formatPseudoElementStyle(className, element, style));
                    clone.appendChild(styleElement);

                    function formatPseudoElementStyle(className, element, style) {
                        var selector = '.' + className + ':' + element;
                        var cssText = style.cssText ? formatCssText(style) : formatCssProperties(style);

                        function formatCssText(style) {
                            var content = style.getPropertyValue('content');
                            return style.cssText + ' content: ' + content + ';';
                        }

                        function formatCssProperties(style) {
                            function formatProperty(name) {
                                return name + ': ' +
                                    style.getPropertyValue(name) +
                                    (style.getPropertyPriority(name) ? ' !important' : '');
                            }

                            return util.asArray(style).map(formatProperty).join('; ') + ';';

                        }
                        return global.document.createTextNode(selector + '{' + cssText + '}');

                    }
                }
            }

            function copyUserInput() {
                if (original instanceof HTMLTextAreaElement) {
                    clone.innerHTML = original.value;
                }
            }

            function fixNamespace() {
                if (clone instanceof SVGElement) {
                    clone.setAttribute('xmlns', 'http://www.w3.org/2000/svg');
                }
            }

            return Promise.resolve(node).then(function(node) {
                return node.cloneNode(false);
            }).then(function(clone) {
                return cloneChildren(node, clone, filter);
            }).then(function(clone) {
                return processClone(node, clone);
            });

        }
        return Promise.resolve().then(cloneStyle).then(clonePseudoElements).then(copyUserInput).then(fixNamespace).then(function() {
            return clone;
        });


    }

    function embedFonts(node) {
        return fontFaces.resolveAll().then(function(cssText) {
            var styleNode = document.createElement('style');
            node.appendChild(styleNode);
            styleNode.appendChild(document.createTextNode(cssText));
            return node;
        });
    }

    function inlineImages(node) {
        return images.inlineAll(node).then(function() {
            return node;
        });
    }

    function makeSvgDataUri(node, width, height) {
        return Promise.resolve(node).then(function(node) {
            node.setAttribute('xmlns', 'http://www.w3.org/1999/xhtml');
            return new XMLSerializer().serializeToString(node);
        }).then(util.escapeXhtml).then(function(xhtml) {
            return '<foreignObject x="0" y="0" width="100%" height="100%">' + xhtml + '</foreignObject>';
        }).then(function(foreignObject) {
            return '<svg xmlns="http://www.w3.org/2000/svg" width="' + width + '" height="' + height + '">' + foreignObject + '</svg>';
        }).then(function(svg) {
            return 'data:image/svg+xml;charset=utf-8,' + svg;
        });
    }

    function draw(domNode, options) {
        function newCanvas(domNode) {
            var canvas = document.createElement('canvas');
            canvas.width = domNode.scrollWidth;
            canvas.height = domNode.scrollHeight;
            return canvas;
        }
        return toSvg(domNode, options).then(util.makeImage).then(util.delay(100)).then(function(image) {
            var canvas = newCanvas(domNode);
            canvas.getContext('2d').drawImage(image, 0, 0);
            return canvas;
        });

    }

    function newUtil() {
        function mimes() {
            /*
             * Only WOFF and EOT mime types for fonts are 'real'
             * see http://www.iana.org/assignments/media-types/media-types.xhtml
             */
            const WOFF = 'application/font-woff';
            const JPEG = 'image/jpeg';

            return {
                'woff': WOFF,
                'woff2': WOFF,
                'ttf': 'application/font-truetype',
                'eot': 'application/vnd.ms-fontobject',
                'png': 'image/png',
                'jpg': JPEG,
                'jpeg': JPEG,
                'gif': 'image/gif',
                'tiff': 'image/tiff',
                'svg': 'image/svg+xml'
            };
        }

        function parseExtension(url) {
            var match = /\.([^\.\/]*?)$/g.exec(url);
            if (match) {
                return match[1];
            } else {
                return '';
            }
        }

        function mimeType(url) {
            var extension = parseExtension(url).toLowerCase();
            return mimes()[extension] || '';
        }

        function isDataUrl(url) {
            return url.search(/^(data:)/) !== -1;
        }

        function toBlob(canvas) {
            return new Promise(function(resolve) {
                var binaryString = window.atob(canvas.toDataURL().split(',')[1]);
                var length = binaryString.length;
                var binaryArray = new Uint8Array(length);

                for (var i = 0; i < length; i++) {
                    binaryArray[i] = binaryString.charCodeAt(i);
                }

                resolve(new Blob([binaryArray], {
                    type: 'image/png'
                }));
            });
        }

        function canvasToBlob(canvas) {
            if (canvas.toBlob) {
                return new Promise(function(resolve) {
                    canvas.toBlob(resolve);
                });
            }
            return toBlob(canvas);
        }

        function resolveUrl(url, baseUrl) {
            var doc = global.document.implementation.createHTMLDocument();
            var base = doc.createElement('base');
            doc.head.appendChild(base);
            var a = doc.createElement('a');
            doc.body.appendChild(a);
            base.href = baseUrl;
            a.href = url;
            return a.href;
        }

        function uid() {
            var index = 0;

            function fourRandomChars() {
                /* see http://stackoverflow.com/a/6248722/2519373 */
                return ('0000' + (Math.random() * Math.pow(36, 4) << 0).toString(36)).slice(-4);
            }

            return function() {
                var result = 'u' + fourRandomChars() + index;
                index++;
                return result;
            };
        }

        function makeImage(uri) {
            return new Promise(function(resolve, reject) {
                var image = new Image();
                image.onload = function() {
                    resolve(image);
                };
                image.onerror = reject;
                image.src = uri;
            });
        }

        function getAndEncode(url) {
            const TIMEOUT = 30000;

            return new Promise(function(resolve, reject) {
                var request = new XMLHttpRequest();

                request.onreadystatechange = done;
                request.ontimeout = timeout;
                request.responseType = 'blob';
                request.timeout = TIMEOUT;
                request.open('GET', url, true);
                request.send();

                function done() {
                    if (request.readyState !== 4) {
                        return;
                    }

                    if (request.status !== 200) {
                        reject(new Error('Cannot fetch resource ' + url + ', status: ' + request.status));
                        return;
                    }

                    var encoder = new FileReader();
                    encoder.onloadend = function() {
                        var content = encoder.result.split(/,/)[1];
                        resolve(content);
                    };
                    encoder.readAsDataURL(request.response);
                }

                function timeout() {
                    reject(new Error('Timeout of ' + TIMEOUT + 'ms occured while fetching resource: ' + url));
                }
            });
        }

        function dataAsUrl(content, type) {
            return 'data:' + type + ';base64,' + content;
        }

        function escape(string) {
            return string.replace(/([.*+?^${}()|\[\]\/\\])/g, '\\$1');
        }

        function delay(ms) {
            return function(arg) {
                return new Promise(function(resolve) {
                    setTimeout(function() {
                        resolve(arg);
                    }, ms);
                });
            };
        }

        function asArray(arrayLike) {
            var array = [];
            var length = arrayLike.length;
            for (var i = 0; i < length; i++) {
                array.push(arrayLike[i]);
            }
            return array;
        }

        function escapeXhtml(string) {
            return string.replace(/#/g, '%23').replace(/\n/g, '%0A');
        }

        return {
            escape: escape,
            parseExtension: parseExtension,
            mimeType: mimeType,
            dataAsUrl: dataAsUrl,
            isDataUrl: isDataUrl,
            canvasToBlob: canvasToBlob,
            resolveUrl: resolveUrl,
            getAndEncode: getAndEncode,
            uid: uid(),
            delay: delay,
            asArray: asArray,
            escapeXhtml: escapeXhtml,
            makeImage: makeImage
        };

    }

    function newInliner() {
        const URL_REGEX = /url\(['"]?([^'"]+?)['"]?\)/g;

        function shouldProcess(string) {
            return string.search(URL_REGEX) !== -1;
        }

        function readUrls(string) {
            var result = [];
            var match;
            while ((match = URL_REGEX.exec(string)) !== null) {
                result.push(match[1]);
            }
            return result.filter(function(url) {
                return !util.isDataUrl(url);
            });
        }

        function inline(string, url, baseUrl, get) {
            function urlAsRegex(url) {
                return new RegExp('(url\\([\'"]?)(' + util.escape(url) + ')([\'"]?\\))', 'g');
            }
            return Promise.resolve(url).then(function(url) {
                return baseUrl ? util.resolveUrl(url, baseUrl) : url;
            }).then(get || util.getAndEncode).then(function(data) {
                return util.dataAsUrl(data, util.mimeType(url));
            }).then(function(dataUrl) {
                return string.replace(urlAsRegex(url), '$1' + dataUrl + '$3');
            });

        }

        function inlineAll(string, baseUrl, get) {
            if (nothingToInline()) {
                return Promise.resolve(string);
            }

            function nothingToInline() {
                return !shouldProcess(string);
            }

            return Promise.resolve(string).then(readUrls).then(function(urls) {
                var done = Promise.resolve(string);
                urls.forEach(function(url) {
                    done = done.then(function(string) {
                        return inline(string, url, baseUrl, get);
                    });
                });
                return done;
            });

        }

        return {
            inlineAll: inlineAll,
            shouldProcess: shouldProcess,
            impl: {
                readUrls: readUrls,
                inline: inline
            }
        };

    }

    function newFontFaces() {
        function resolveAll() {
            return readAll(document).then(function(webFonts) {
                return Promise.all(
                    webFonts.map(function(webFont) {
                        return webFont.resolve();
                    }));
            }).then(function(cssStrings) {
                return cssStrings.join('\n');
            });
        }

        function readAll() {
            function selectWebFontRules(cssRules) {
                return cssRules.filter(function(rule) {
                    return rule.type === CSSRule.FONT_FACE_RULE;
                }).filter(function(rule) {
                    return inliner.shouldProcess(rule.style.getPropertyValue('src'));
                });
            }

            function getCssRules(styleSheets) {
                var cssRules = [];
                styleSheets.forEach(function(sheet) {
                    try {
                        util.asArray(sheet.cssRules || []).forEach(cssRules.push.bind(cssRules));
                    } catch (e) {
                        console.log('Error while reading CSS rules from ' + sheet.href, e.toString());
                    }
                });
                return cssRules;
            }

            function newWebFont(webFontRule) {
                return {
                    resolve: function resolve() {
                        var baseUrl = (webFontRule.parentStyleSheet || {}).href;
                        return inliner.inlineAll(webFontRule.cssText, baseUrl);
                    },
                    src: function() {
                        return webFontRule.style.getPropertyValue('src');
                    }
                };
            }
            return Promise.resolve(util.asArray(document.styleSheets)).then(getCssRules).then(selectWebFontRules).then(function(rules) {
                return rules.map(newWebFont);
            });


        }

        return {
            resolveAll: resolveAll,
            impl: {
                readAll: readAll
            }
        };

    }

    function newImages() {
        function newImage(element) {
            function inline(get) {
                if (util.isDataUrl(element.src)) {
                    return Promise.resolve();
                }

                return Promise.resolve(element.src).then(get || util.getAndEncode).then(function(data) {
                    return util.dataAsUrl(data, util.mimeType(element.src));
                }).then(function(dataUrl) {
                    return new Promise(function(resolve, reject) {
                        element.onload = resolve;
                        element.onerror = reject;
                        element.src = dataUrl;
                    });
                });
            }
            return {
                inline: inline
            };

        }

        function inlineAll(node) {
            if (!(node instanceof Element)) {
                return Promise.resolve(node);
            }

            function inlineBackground(node) {
                var background = node.style.getPropertyValue('background');

                if (!background) {
                    return Promise.resolve(node);
                }

                return inliner.inlineAll(background).then(function(inlined) {
                    node.style.setProperty(
                        'background',
                        inlined,
                        node.style.getPropertyPriority('background'));
                }).then(function() {
                    return node;
                });
            }

            return inlineBackground(node).then(function() {
                if (node instanceof HTMLImageElement) {
                    return newImage(node).inline();
                } else {
                    return Promise.all(
                        util.asArray(node.childNodes).map(function(child) {
                            return inlineAll(child);
                        }));
                }
            });

        }
        return {
            inlineAll: inlineAll,
            impl: {
                newImage: newImage
            }
        };

    }
})(this);

**Comment by tsayen** [comment]
OK, just to be clear, I use JSHint for static analysis http://jshint.com/about/

**Comment by vcrepin** [comment]
Yes, To my knowledge Lint is much more restrictive. It reported numerous warnings in the original version (most of them are related to the ordering of the functions) that I corrected. But mostly it spotted an error in the regexp that I think was breaking the script in safari. More on this later tonight...

**Comment by vcrepin** [comment]
The problem was const:

error    SyntaxError: Unexpected keyword 'const'. Const declarations are not supported in strict mode., http://localhost:8100/lib/domtoimage/dom-to-image.js, Line: 237

**Comment by tsayen** [comment]
Hmm, it turns out that `const` support in Safari is [strange](http://caniuse.com/#feat=const) ("only recognized when not in strict mode"), makes me consider replacing all the `const`s with `var`

**Comment by kapyaar** [comment]
@vcrepin I am having issues with mobile browser. Can you please check this fiddle on chrome? or safari? and see if you can get it working? It works for me on desktop versions. 

http://jsfiddle.net/8h13tmcj/25/

**Comment by kircof** [comment]
@kapyaar Hi, I tested your fiddle and it works on chrome but I get the "oops, something went wrong!" alert. I did the suggested solution in #40 and it works fine but the resultant image is svg and I cant transform it to png because i get the "SecurityError: DOM Exception 18: An attempt was made to break through the security policy of the user agent." Im still trying and will keep an eye on new suggestions. Hope this helps.
